### PR TITLE
Rework token prices

### LINKED ISSUE #6547: Address page -> tokens tab: sort items by value DESC, name ASC by default
Currently, token balances are sorted by the amount in descending order.

### LINKED ISSUE #6855: Address tokens pagination issue
Now we paging by `token_name`, `token_type` and `value`. That's not sufficent when address has a lot of instances of `ERC-1155`(with the same amount) or `ERC-721` tokens.

### PR DESCRIPTION
**Pull request by sl1depengwyn**
Close #6547 
Close #6855 

## Motivation

It happened that now we have no token price feature in master branch. Also, we sorted token balances by value that is not very useful for user. Additionally, we had issue of paginating balances of ERC-1155 or ERC-721 tokens when user have the same amount of different token ids of the same token 

## Changelog

To be able to sort and paginate token balances I implemented new token balance fetcher, changed pagination params for token balances, removed redundant functionality like manual sorting of token balance after fetching from the database since now the are fetching already in right order. Also, I removed `KnownTokens` module and functionality as new token price fetched does not use it.

Docs update https://github.com/blockscout/docs/pull/117

Address page tokens before:
image
after:
image

Top tokens page before:
image

after:
image

## Checklist for your Pull Request (PR)

  - [x] I added an entry to `CHANGELOG.md` with this PR
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [ ] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [x] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. https://github.com/blockscout/docs/pull/117
  - [ ] If I add new indices into DB, I checked, that they are not redundant with PGHero or other tools

### CODE REVIEW

File: config/runtime.exs [hunk 1]
```diff
@@ -273,7 +270,41 @@ exchange_rates_source =
...
+config :explorer, Explorer.ExchangeRates.Source.CoinMarketCap,
+  api_key: System.get_env("EXCHANGE_RATES_COINMARKETCAP_API_KEY")
+
+config :explorer, Explorer.ExchangeRates.Source.CoinGecko,
+  platfortm: System.get_env("EXCHANGE_RATES_COINGECKO_PLATFORM_ID"),
+  api_key: System.get_env("EXCHANGE_RATES_COINGECKO_API_KEY"),
+  coin_id: System.get_env("EXCHANGE_RATES_COINGECKO_COIN_ID")
+
+token_excange_rate_interval =
+  case "TOKEN_EXCANGE_RATE_INTERVAL" |> System.get_env("") |> String.downcase() |> Integer.parse() do
```
Comment (vbaranov) [inline config/runtime.exs:282]: @sl1depengwyn Typo, missing "h". Also, please fix in the docs https://github.com/blockscout/docs/pull/117

File: config/runtime.exs [hunk 2]
```diff
@@ -273,7 +270,41 @@ exchange_rates_source =
...
+    {hours, "h"} -> :timer.hours(hours)
+    {minutes, "m"} -> :timer.minutes(minutes)
+    {seconds, s} when s in ["s", ""] -> :timer.seconds(seconds)
+    _ -> nil
+  end
+
+token_excange_rate_refetch_interval =
+  case "TOKEN_EXCANGE_RATE_REFETCH_INTERVAL" |> System.get_env("") |> String.downcase() |> Integer.parse() do
```
Comment (vbaranov) [inline config/runtime.exs:290]: Typo

File: config/runtime.exs [hunk 3]
```diff
@@ -273,7 +270,41 @@ exchange_rates_source =
...
+    {hours, h} when h in ["h", ""] -> :timer.hours(hours)
+    {minutes, "m"} -> :timer.minutes(minutes)
+    {seconds, "s"} -> :timer.seconds(seconds)
+    _ -> nil
+  end
+
+token_exchange_rate_max_batch_size =
+  case "TOKEN_EXCANGE_RATE_MAX_BATCH_SIZE" |> System.get_env("") |> Integer.parse() do
```
Comment (vbaranov) [inline config/runtime.exs:298]: Typo

File: config/runtime.exs [hunk 4]
```diff
@@ -273,7 +270,41 @@ exchange_rates_source =
...
+    {batch_size, ""} -> batch_size
+    _ -> nil
+  end
+
+config :explorer, Explorer.ExchangeRates.TokenExchangeRates,
+  enabled: System.get_env("DISABLE_TOKEN_EXCANGE_RATE") != "true",
```
Comment (vbaranov) [inline config/runtime.exs:304]: Typo

**Review by vbaranov — COMMENTED** [review]
@sl1depengwyn does it make sense to display price on "top tokens page" `/tokens` as well and make the same default sorting?

**Comment by sl1depengwyn** [comment]
> @sl1depengwyn does it make sense to display price on top tokens page `/tokens` as well and make the same default sorting?

You mean sort by total supply * token price? In this case I think it make sense

**Comment by vbaranov** [comment]
> > @sl1depengwyn does it make sense to display price on top tokens page `/tokens` as well and make the same default sorting?
> 
> You mean sort by total supply * token price? In this case I think it make sense

yes

**Review by vbaranov — CHANGES_REQUESTED** [review]
/verified-contracts page doesn't open with the changes

```
Request: GET /verified-contracts?type=JSON
** (exit) an exception was raised:
    ** (KeyError) key :market_cap not found in: %Explorer.Chain.Token{__meta__: #Ecto.Schema.Metadata<:loaded, "tokens">, name: "Seizon", symbol: "Seizon", total_supply: #Decimal<7573>, decimals: nil, type: "ERC-721", cataloged: true, holder_count: 3362, skip_metadata: nil, total_supply_updated_at_block: nil, fiat_value: nil, circulating_market_cap: nil, contract_address_hash: %Explorer.Chain.Hash{byte_count: 20, bytes: <<166, 205, 39, 40, 116, 238, 124, 135, 46, 182, 104, 1, 239, 246, 39, 132, 192, 177, 50, 133>>}, contract_address: #Ecto.Association.NotLoaded<association :contract_address is not loaded>, inserted_at: ~U[2023-03-06 23:59:07.423871Z], updated_at: ~U[2023-03-08 11:53:03.304706Z]}
        (block_scout_web 5.1.1) lib/block_scout_web/templates/verified_contracts/_contract.html.eex:58: BlockScoutWeb.VerifiedContractsView."_contract.html"/1
        (phoenix 1.5.13) lib/phoenix/view.ex:472: Phoenix.View.render_to_iodata/3
        (phoenix 1.5.13) lib/phoenix/view.ex:479: Phoenix.View.render_to_string/3
        (block_scout_web 5.1.1) lib/block_scout_web/controllers/verified_contracts_controller.ex:27: anonymous fn/2 in BlockScoutWeb.VerifiedContractsController.index/2
        (elixir 1.14.0) lib/enum.ex:2468: Enum."-reduce/3-lists^foldl/2-0-"/3
        (block_scout_web 5.1.1) lib/block_scout_web/controllers/verified_contracts_controller.ex:26: BlockScoutWeb.VerifiedContractsController.index/2
        (block_scout_web 5.1.1) lib/block_scout_web/controllers/verified_contracts_controller.ex:1: BlockScoutWeb.VerifiedContractsController.action/2
        (block_scout_web 5.1.1) lib/block_scout_web/controllers/verified_contracts_controller.ex:1: BlockScoutWeb.VerifiedContractsController.phoenix_controller_pipeline/2
```

**Comment by sl1depengwyn** [comment]
> /verified-contracts page doesn't open with the changes
> 
> ```
> Request: GET /verified-contracts?type=JSON
> ** (exit) an exception was raised:
>     ** (KeyError) key :market_cap not found in: %Explorer.Chain.Token{__meta__: #Ecto.Schema.Metadata<:loaded, "tokens">, name: "Seizon", symbol: "Seizon", total_supply: #Decimal<7573>, decimals: nil, type: "ERC-721", cataloged: true, holder_count: 3362, skip_metadata: nil, total_supply_updated_at_block: nil, fiat_value: nil, circulating_market_cap: nil, contract_address_hash: %Explorer.Chain.Hash{byte_count: 20, bytes: <<166, 205, 39, 40, 116, 238, 124, 135, 46, 182, 104, 1, 239, 246, 39, 132, 192, 177, 50, 133>>}, contract_address: #Ecto.Association.NotLoaded<association :contract_address is not loaded>, inserted_at: ~U[2023-03-06 23:59:07.423871Z], updated_at: ~U[2023-03-08 11:53:03.304706Z]}
>         (block_scout_web 5.1.1) lib/block_scout_web/templates/verified_contracts/_contract.html.eex:58: BlockScoutWeb.VerifiedContractsView."_contract.html"/1
>         (phoenix 1.5.13) lib/phoenix/view.ex:472: Phoenix.View.render_to_iodata/3
>         (phoenix 1.5.13) lib/phoenix/view.ex:479: Phoenix.View.render_to_string/3
>         (block_scout_web 5.1.1) lib/block_scout_web/controllers/verified_contracts_controller.ex:27: anonymous fn/2 in BlockScoutWeb.VerifiedContractsController.index/2
>         (elixir 1.14.0) lib/enum.ex:2468: Enum."-reduce/3-lists^foldl/2-0-"/3
>         (block_scout_web 5.1.1) lib/block_scout_web/controllers/verified_contracts_controller.ex:26: BlockScoutWeb.VerifiedContractsController.index/2
>         (block_scout_web 5.1.1) lib/block_scout_web/controllers/verified_contracts_controller.ex:1: BlockScoutWeb.VerifiedContractsController.action/2
>         (block_scout_web 5.1.1) lib/block_scout_web/controllers/verified_contracts_controller.ex:1: BlockScoutWeb.VerifiedContractsController.phoenix_controller_pipeline/2
> ```

Should be fixed in https://github.com/blockscout/blockscout/pull/6925/commits/adb5d7a23b7114e728628e04aad5c77b8605b980

File: apps/block_scout_web/assets/js/pages/token/search.js [hunk 5]
```diff
@@ -24,12 +26,26 @@ export function reducer (state, action) {
...
 }
 
+const elements = {
+  '[data-usd-value]': {
+    render ($el, state) {
+      $el.each((i, el) => {
+        el.innerHTML = formatUsdValue(el.dataset.usdValue)
+      })
+      // @ts-ignore
+      if (state.channelDisconnected && !window.loading) $el.show()
```
Comment (nikitosing) [inline apps/block_scout_web/assets/js/pages/token/search.js:36]: Why `$el.show()` depends on `channelDisconnected`?

File: config/runtime.exs [hunk 6]
```diff
@@ -328,7 +325,42 @@ exchange_rates_source =
 
 config :explorer, Explorer.ExchangeRates.Source, source: exchange_rates_source
 
-config :explorer, Explorer.KnownTokens, enabled: System.get_env("DISABLE_KNOWN_TOKENS") != "true", store: :ets
```
Comment (nikitosing) [inline config/runtime.exs:331]: Let's delete this env also from `docker-compose/envs/common-blockscout.env` and `docker/Makefile`

File: apps/explorer/lib/explorer/exchange_rates/source/coin_gecko.ex [hunk 7]
```diff
@@ -177,6 +209,14 @@ defmodule Explorer.ExchangeRates.Source.CoinGecko do
     end
   end
 
+  defp platform do
+    config(:platform) || "ethereum"
+  end
+
+  defp currency do
+    config(:currency) || "usd"
```
Comment (nikitosing) [inline apps/explorer/lib/explorer/exchange_rates/source/coin_gecko.ex:217]: For now we haven't `:currency` in config. Is it for future improvements?

File: apps/explorer/lib/explorer/chain.ex [hunk 8]
```diff
@@ -4712,12 +4723,52 @@ defmodule Explorer.Chain do
...
 
-  def page_current_token_balances(query, paging_options: %PagingOptions{key: {name, type, value}}) do
+  def page_current_token_balances(query, paging_options: %PagingOptions{key: {nil, nil, type, id}}) do
+    fiat_balance = CurrentTokenBalance.fiat_value_query()
+
+    condition =
+      dynamic(
+        [ctb, t],
+        is_nil(^fiat_balance) and
+          (t.type > ^type or
```
Comment (nikitosing) [inline apps/explorer/lib/explorer/chain.ex:4733]: Why did you decide to remove ordering by value?

File: apps/explorer/lib/explorer/exchange_rates/source/coin_gecko.ex (hunk 7 above)
Comment (sl1depengwyn) [reply apps/explorer/lib/explorer/exchange_rates/source/coin_gecko.ex:217]: Yes, it is

File: apps/block_scout_web/assets/js/pages/token/search.js (hunk 5 above)
Comment (sl1depengwyn) [reply apps/block_scout_web/assets/js/pages/token/search.js]: Fixed in https://github.com/blockscout/blockscout/pull/6925/commits/f0144911ab944205c2ea9606ee3a1aaa8dcedd6e

File: config/runtime.exs (hunk 6 above)
Comment (sl1depengwyn) [reply config/runtime.exs:334]: Fixed in https://github.com/blockscout/blockscout/pull/6925/commits/f0144911ab944205c2ea9606ee3a1aaa8dcedd6e

File: apps/block_scout_web/assets/js/pages/verified_contracts.js [hunk 9]
```diff
@@ -35,6 +36,15 @@ const elements = {
...
       return $el.show()
     }
+  },
+  '[data-usd-value]': {
+    render ($el, state) {
+      $el.each((i, el) => {
+        el.innerHTML = formatUsdValue(el.dataset.usdValue)
+      })
+      // @ts-ignore
+      if (state.channelDisconnected && !window.loading) $el.show()
```
[the remark this answers is not in the document]
Comment (sl1depengwyn) [reply apps/block_scout_web/assets/js/pages/verified_contracts.js]: Fixed in https://github.com/blockscout/blockscout/pull/6925/commits/f0144911ab944205c2ea9606ee3a1aaa8dcedd6e

**Review by vbaranov — APPROVED** [review]

File: apps/explorer/lib/explorer/chain.ex [hunk 10]
```diff
@@ -4728,12 +4740,33 @@ defmodule Explorer.Chain do
...
+      query,
+      [ctb, t],
+      ^condition
+    )
+  end
+
+  def page_current_token_balances(query, %PagingOptions{key: {fiat_value, _value, _id}}) do
+    fiat_balance = CurrentTokenBalance.fiat_value_query()
+
+    condition = dynamic([ctb, t], ^fiat_balance < ^fiat_value or is_nil(^fiat_balance))
```
Comment (nikitosing) [inline apps/explorer/lib/explorer/chain.ex:4764]: Shouldn't we take into account case when some current token balances can have the same fiat value?

Comment (sl1depengwyn) [reply apps/explorer/lib/explorer/chain.ex]: Nice catch, fixed in https://github.com/blockscout/blockscout/pull/6925/commits/061f99232d1c0ef15f34f2385ec4eab70fde78cf#diff-3fce0f082a5cb927aa738ef8f7d582824df91d08957971d36a0873efacd88816

**Review by nikitosing — APPROVED** [review]

**Pull request merged**